weave/trace_server/errors.py

```diff
@@ -39,15 +76,227 @@ def __init__(self, message: str, api_key_name: str):
+    global _error_registry
+    if _error_registry is None:
+        _error_registry = ErrorRegistry()
+    return _error_registry
+
+
+def handle_server_exception(exc: Exception) -> tuple[dict[str, Any], int]:
+    """Handle a server exception."""
+    registry = _get_error_registry()
+    return registry.handle_exception(exc)
+
+
+def get_registered_error_classes() -> list[type[Exception]]:
+    """Get all registered error classes."""
+    registry = _get_error_registry()
+    return list(registry.get_all_definitions().keys())
+
+
+def handle_clickhouse_query_error(e: Exception) -> None:
+    """
+    Handle common ClickHouse query errors by raising appropriate custom exceptions.
+
+    Args:
+        e: The original exception from ClickHouse
+
+    Raises:
+        QueryMemoryLimitExceeded: When the query exceeds memory limits
+        QueryNoCommonType: When there's a type mismatch in the query
+        QueryTimeoutExceeded: When the query exceeds timeout limits
+        Exception: Re-raises the original exception if no known pattern matches
+    """
+    error_str = str(e)
+
+    limit_scope_message = (
+        "Please limit the scope of the query by including a date range and/or additional filter"
+        " criteria. "
+    )
+
+    if "MEMORY_LIMIT_EXCEEDED" in error_str:
+        raise QueryMemoryLimitExceeded(
+            "Query memory limit exceeded. " + limit_scope_message
+        ) from e
+    if "NO_COMMON_TYPE" in error_str:
+        raise QueryNoCommonType(
+            "No common type between data types in query. "
+            "This can occur when comparing types without using the $convert operation. "
+            "Example: filtering calls by inputs.integer_value = 1 without using $convert -> "
+            "Correct: {$expr: {$eq: [{$convert: {input: {$getField: 'inputs.integer_value'}, to: 'double'}}, {$literal: 1}]}}"
+        ) from e
+    if "TIMEOUT_EXCEEDED" in error_str:
+        raise QueryTimeoutExceeded(
+            "Query timeout exceeded. " + limit_scope_message
+        ) from e
+
+    # Re-raise the original exception if no known pattern matches
+    raise
+
+
+def _format_transport_query_error(exc: Exception) -> dict[str, Any]:
+    """Format TransportQueryError with special permission logic."""
+    transport_exc = exc if isinstance(exc, TransportQueryError) else None
+    if transport_exc and transport_exc.errors:
+        for error in transport_exc.errors:
+            if error.get("extensions", {}).get("code") == "PERMISSION_ERROR":
+                return {"reason": "Forbidden"}
+            if error.get("message") == "project not found" and error.get("path") == [
+                "upsertBucket"
+            ]:
+                # This seems counter intuitive, but gorilla returns this error when the project exists
+                # but the user does not have access to it. This seems like a bug on Gorilla's side.
+                return {"reason": "Forbidden"}
+    return {"reason": "Forbidden"}
```

Comment:
I know this is not a new function in the PR, but this seems like useless code since it always returns `{"reason": "Forbidden"}`. Let's yank it.